// @angular/material/datepicker/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { MdMonthView } from './month-view';
import { CommonModule } from '@angular/common';
import { A11yModule, OverlayModule, StyleModule } from '../core';
import { MdCalendarBody } from './calendar-body';
import { MdYearView } from './year-view';
import { MdDatepicker, MdDatepickerContent, MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER, } from './datepicker';
import { MdDatepickerInput } from './datepicker-input';
import { MdDialogModule } from '../dialog/index';
import { MdCalendar } from './calendar';
import { MdDatepickerToggle } from './datepicker-toggle';
import { MdButtonModule } from '../button/index';
import { MdDatepickerIntl } from './datepicker-intl';
import { MdIconModule } from '../icon/index';
export { MdCalendar } from './calendar';
export { MdCalendarCell, MdCalendarBody } from './calendar-body';
export { MD_DATEPICKER_SCROLL_STRATEGY, MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER_FACTORY, MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER, MdDatepickerContent, MdDatepicker } from './datepicker';
export { MD_DATEPICKER_VALUE_ACCESSOR, MD_DATEPICKER_VALIDATORS, MdDatepickerInputEvent, MdDatepickerInput } from './datepicker-input';
export { MdDatepickerIntl } from './datepicker-intl';
export { MdDatepickerToggle } from './datepicker-toggle';
export { MdMonthView } from './month-view';
export { MdYearView } from './year-view';
var MdDatepickerModule = (function () {
    function MdDatepickerModule() {
    }
    return MdDatepickerModule;
}());
export { MdDatepickerModule };
MdDatepickerModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    MdButtonModule,
                    MdDialogModule,
                    MdIconModule,
                    OverlayModule,
                    StyleModule,
                    A11yModule,
                ],
                exports: [
                    MdDatepicker,
                    MdDatepickerContent,
                    MdDatepickerInput,
                    MdDatepickerToggle,
                ],
                declarations: [
                    MdCalendar,
                    MdCalendarBody,
                    MdDatepicker,
                    MdDatepickerContent,
                    MdDatepickerInput,
                    MdDatepickerToggle,
                    MdMonthView,
                    MdYearView,
                ],
                providers: [
                    MdDatepickerIntl,
                    MD_DATEPICKER_SCROLL_STRATEGY_PROVIDER,
                ],
                entryComponents: [
                    MdDatepickerContent,
                ]
            },] },
];
/**
 * @nocollapse
 */
MdDatepickerModule.ctorParameters = function () { return []; };
function MdDatepickerModule_tsickle_Closure_declarations() {
    /** @type {?} */
    MdDatepickerModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    MdDatepickerModule.ctorParameters;
}
