// @angular/material/core/datetime/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { DateAdapter } from './date-adapter';
import { NativeDateAdapter } from './native-date-adapter';
import { MD_DATE_FORMATS } from './date-formats';
import { MD_NATIVE_DATE_FORMATS } from './native-date-formats';
export { DateAdapter } from './date-adapter';
export { MD_DATE_FORMATS } from './date-formats';
export { NativeDateAdapter } from './native-date-adapter';
export { MD_NATIVE_DATE_FORMATS } from './native-date-formats';
var NativeDateModule = (function () {
    function NativeDateModule() {
    }
    return NativeDateModule;
}());
export { NativeDateModule };
NativeDateModule.decorators = [
    { type: NgModule, args: [{
                providers: [{ provide: DateAdapter, useClass: NativeDateAdapter }],
            },] },
];
/**
 * @nocollapse
 */
NativeDateModule.ctorParameters = function () { return []; };
function NativeDateModule_tsickle_Closure_declarations() {
    /** @type {?} */
    NativeDateModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    NativeDateModule.ctorParameters;
}
var MdNativeDateModule = (function () {
    function MdNativeDateModule() {
    }
    return MdNativeDateModule;
}());
export { MdNativeDateModule };
MdNativeDateModule.decorators = [
    { type: NgModule, args: [{
                imports: [NativeDateModule],
                providers: [{ provide: MD_DATE_FORMATS, useValue: MD_NATIVE_DATE_FORMATS }],
            },] },
];
/**
 * @nocollapse
 */
MdNativeDateModule.ctorParameters = function () { return []; };
function MdNativeDateModule_tsickle_Closure_declarations() {
    /** @type {?} */
    MdNativeDateModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    MdNativeDateModule.ctorParameters;
}
