// @angular/material/chips/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { MdChipList } from './chip-list';
import { MdBasicChip, MdChip, MdChipRemove } from './chip';
import { MdChipInput } from './chip-input';
export { MdChipList } from './chip-list';
export { MdChipBase, _MdChipMixinBase, MdBasicChip, MdChip, MdChipRemove } from './chip';
export { MdChipInput } from './chip-input';
var MdChipsModule = (function () {
    function MdChipsModule() {
    }
    return MdChipsModule;
}());
export { MdChipsModule };
MdChipsModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                exports: [MdChipList, MdChip, MdChipInput, MdChipRemove, MdChipRemove, MdBasicChip],
                declarations: [MdChipList, MdChip, MdChipInput, MdChipRemove, MdChipRemove, MdBasicChip]
            },] },
];
/**
 * @nocollapse
 */
MdChipsModule.ctorParameters = function () { return []; };
function MdChipsModule_tsickle_Closure_declarations() {
    /** @type {?} */
    MdChipsModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    MdChipsModule.ctorParameters;
}
