// @angular/material/dialog/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OverlayModule, PortalModule, A11yModule, MdCommonModule, } from '../core';
import { MdDialog, MD_DIALOG_SCROLL_STRATEGY_PROVIDER } from './dialog';
import { MdDialogContainer } from './dialog-container';
import { MdDialogClose, MdDialogContent, MdDialogTitle, MdDialogActions } from './dialog-content-directives';
var MdDialogModule = (function () {
    function MdDialogModule() {
    }
    return MdDialogModule;
}());
export { MdDialogModule };
MdDialogModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    OverlayModule,
                    PortalModule,
                    A11yModule,
                    MdCommonModule,
                ],
                exports: [
                    MdDialogContainer,
                    MdDialogClose,
                    MdDialogTitle,
                    MdDialogContent,
                    MdDialogActions,
                    MdCommonModule,
                ],
                declarations: [
                    MdDialogContainer,
                    MdDialogClose,
                    MdDialogTitle,
                    MdDialogActions,
                    MdDialogContent,
                ],
                providers: [
                    MdDialog,
                    MD_DIALOG_SCROLL_STRATEGY_PROVIDER,
                ],
                entryComponents: [MdDialogContainer],
            },] },
];
/**
 * @nocollapse
 */
MdDialogModule.ctorParameters = function () { return []; };
function MdDialogModule_tsickle_Closure_declarations() {
    /** @type {?} */
    MdDialogModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    MdDialogModule.ctorParameters;
}
export { MD_DIALOG_DATA, MD_DIALOG_SCROLL_STRATEGY, MD_DIALOG_SCROLL_STRATEGY_PROVIDER_FACTORY, MD_DIALOG_SCROLL_STRATEGY_PROVIDER, MdDialog } from './dialog';
export { throwMdDialogContentAlreadyAttachedError, MdDialogContainer } from './dialog-container';
export { MdDialogClose, MdDialogTitle, MdDialogContent, MdDialogActions } from './dialog-content-directives';
export { MdDialogConfig } from './dialog-config';
export { MdDialogRef } from './dialog-ref';
