// dgeni-packages/typescript/services/convertPrivateClassesToInterfaces.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ClassExportDoc_1 = require("../api-doc-types/ClassExportDoc");
var FileInfo_1 = require("./TsParser/FileInfo");
var convertPrivateClassesToInterfaces_1 = require("./convertPrivateClassesToInterfaces");
describe('convertPrivateClassesToInterfaces', function () {
    var basePath = 'a/b/c';
    var moduleDoc = { id: 'someModule' };
    var mockDeclaration = { getSourceFile: function () { return ({ fileName: 'x/y/z', text: 'blah blah' }); } };
    var classSymbol = {
        getDeclarations: function () { return [mockDeclaration]; },
        name: 'privateClass',
        valueDeclaration: mockDeclaration,
    };
    var classDoc;
    var docs;
    beforeEach(function () {
        spyOn(FileInfo_1.FileInfo.prototype, 'getRealFilePath').and.callFake(function (filePath) { return filePath; });
        classDoc = new ClassExportDoc_1.ClassExportDoc(moduleDoc, classSymbol, basePath, true, []);
        classDoc.constructorDoc = { internal: true };
        docs = [classDoc];
    });
    it('should convert private class docs to interface docs', function () {
        convertPrivateClassesToInterfaces_1.convertPrivateClassesToInterfaces(docs, false);
        expect(docs[0].docType).toEqual('interface');
    });
    it('should not touch non-private class docs', function () {
        classDoc.constructorDoc = {};
        convertPrivateClassesToInterfaces_1.convertPrivateClassesToInterfaces(docs, false);
        expect(docs[0].docType).toEqual('class');
    });
    it('should convert the heritage since interfaces use `extends` not `implements`', function () {
        classDoc.implementsClauses = ['parentInterface'];
        convertPrivateClassesToInterfaces_1.convertPrivateClassesToInterfaces(docs, false);
        expect(docs[0].extendsClauses).toEqual(['parentInterface']);
    });
    it('should add new injectable reference types, if specified, to the passed in collection', function () {
        convertPrivateClassesToInterfaces_1.convertPrivateClassesToInterfaces(docs, true);
        expect(docs[1]).toEqual(jasmine.objectContaining({
            docType: 'const',
            id: 'someModule/privateClass',
            name: 'privateClass',
            type: 'InjectableReference',
        }));
    });
});
