// @angular/material/button-toggle/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { MdButtonToggleGroup, MdButtonToggleGroupMultiple, MdButtonToggle } from './button-toggle';
import { UNIQUE_SELECTION_DISPATCHER_PROVIDER, MdCommonModule, StyleModule, } from '../core';
var MdButtonToggleModule = (function () {
    function MdButtonToggleModule() {
    }
    return MdButtonToggleModule;
}());
export { MdButtonToggleModule };
MdButtonToggleModule.decorators = [
    { type: NgModule, args: [{
                imports: [MdCommonModule, StyleModule],
                exports: [
                    MdButtonToggleGroup,
                    MdButtonToggleGroupMultiple,
                    MdButtonToggle,
                    MdCommonModule,
                ],
                declarations: [MdButtonToggleGroup, MdButtonToggleGroupMultiple, MdButtonToggle],
                providers: [UNIQUE_SELECTION_DISPATCHER_PROVIDER]
            },] },
];
/**
 * @nocollapse
 */
MdButtonToggleModule.ctorParameters = function () { return []; };
function MdButtonToggleModule_tsickle_Closure_declarations() {
    /** @type {?} */
    MdButtonToggleModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    MdButtonToggleModule.ctorParameters;
}
export { MdButtonToggleGroupBase, _MdButtonToggleGroupMixinBase, MD_BUTTON_TOGGLE_GROUP_VALUE_ACCESSOR, MdButtonToggleChange, MdButtonToggleGroup, MdButtonToggleGroupMultiple, MdButtonToggle } from './button-toggle';
