// @angular/material/core/style/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { CdkMonitorFocus, FOCUS_ORIGIN_MONITOR_PROVIDER } from './focus-origin-monitor';
import { PlatformModule } from '../platform/index';
var StyleModule = (function () {
    function StyleModule() {
    }
    return StyleModule;
}());
export { StyleModule };
StyleModule.decorators = [
    { type: NgModule, args: [{
                imports: [PlatformModule],
                declarations: [CdkMonitorFocus],
                exports: [CdkMonitorFocus],
                providers: [FOCUS_ORIGIN_MONITOR_PROVIDER],
            },] },
];
/**
 * @nocollapse
 */
StyleModule.ctorParameters = function () { return []; };
function StyleModule_tsickle_Closure_declarations() {
    /** @type {?} */
    StyleModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    StyleModule.ctorParameters;
}
export { TOUCH_BUFFER_MS, FocusOriginMonitor, CdkMonitorFocus, FOCUS_ORIGIN_MONITOR_PROVIDER_FACTORY, FOCUS_ORIGIN_MONITOR_PROVIDER } from './focus-origin-monitor';
export { applyCssTransform } from './apply-transform';
