// @angular/material/table/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { MdTable } from './table';
import { CdkTableModule } from '@angular/cdk/table';
import { MdCell, MdHeaderCell, MdCellDef, MdHeaderCellDef, MdColumnDef } from './cell';
import { MdHeaderRow, MdRow, MdHeaderRowDef, MdRowDef } from './row';
import { CommonModule } from '@angular/common';
import { MdCommonModule } from '../core';
export { _MdCellDef, _MdHeaderCellDef, _MdColumnDef, _MdHeaderCell, _MdCell, MdCellDef, MdHeaderCellDef, MdColumnDef, MdHeaderCell, MdCell } from './cell';
export { _MdTable, MdTable } from './table';
export { _MdHeaderRowDef, _MdCdkRowDef, _MdHeaderRow, _MdRow, MdHeaderRowDef, MdRowDef, MdHeaderRow, MdRow } from './row';
var MdTableModule = (function () {
    function MdTableModule() {
    }
    return MdTableModule;
}());
export { MdTableModule };
MdTableModule.decorators = [
    { type: NgModule, args: [{
                imports: [CdkTableModule, CommonModule, MdCommonModule],
                exports: [MdTable, MdCellDef, MdHeaderCellDef, MdColumnDef,
                    MdHeaderRowDef, MdRowDef,
                    MdHeaderCell, MdCell, MdHeaderRow, MdRow],
                declarations: [MdTable, MdCellDef, MdHeaderCellDef, MdColumnDef,
                    MdHeaderRowDef, MdRowDef,
                    MdHeaderCell, MdCell, MdHeaderRow, MdRow],
            },] },
];
/**
 * @nocollapse
 */
MdTableModule.ctorParameters = function () { return []; };
function MdTableModule_tsickle_Closure_declarations() {
    /** @type {?} */
    MdTableModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    MdTableModule.ctorParameters;
}
