// @angular/material/radio/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VIEWPORT_RULER_PROVIDER } from '../core/overlay/position/viewport-ruler';
import { MdRippleModule, MdCommonModule, UNIQUE_SELECTION_DISPATCHER_PROVIDER, FocusOriginMonitor, } from '../core';
import { MdRadioGroup, MdRadioButton } from './radio';
var MdRadioModule = (function () {
    function MdRadioModule() {
    }
    return MdRadioModule;
}());
export { MdRadioModule };
MdRadioModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, MdRippleModule, MdCommonModule],
                exports: [MdRadioGroup, MdRadioButton, MdCommonModule],
                providers: [UNIQUE_SELECTION_DISPATCHER_PROVIDER, VIEWPORT_RULER_PROVIDER, FocusOriginMonitor],
                declarations: [MdRadioGroup, MdRadioButton],
            },] },
];
/**
 * @nocollapse
 */
MdRadioModule.ctorParameters = function () { return []; };
function MdRadioModule_tsickle_Closure_declarations() {
    /** @type {?} */
    MdRadioModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    MdRadioModule.ctorParameters;
}
export { MD_RADIO_GROUP_CONTROL_VALUE_ACCESSOR, MdRadioChange, MdRadioGroupBase, _MdRadioGroupMixinBase, MdRadioGroup, MdRadioButtonBase, _MdRadioButtonMixinBase, MdRadioButton } from './radio';
