// @angular/material/core/ripple/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { MdRipple } from './ripple';
import { MdCommonModule } from '../common-behaviors/common-module';
import { VIEWPORT_RULER_PROVIDER } from '../overlay/position/viewport-ruler';
import { ScrollDispatchModule } from '../overlay/scroll/index';
import { PlatformModule } from '../platform/index';
export { MdRipple, MD_RIPPLE_GLOBAL_OPTIONS } from './ripple';
export { RippleRef, RippleState } from './ripple-ref';
export { RIPPLE_FADE_IN_DURATION, RIPPLE_FADE_OUT_DURATION } from './ripple-renderer';
var MdRippleModule = (function () {
    function MdRippleModule() {
    }
    return MdRippleModule;
}());
export { MdRippleModule };
MdRippleModule.decorators = [
    { type: NgModule, args: [{
                imports: [MdCommonModule, PlatformModule, ScrollDispatchModule],
                exports: [MdRipple, MdCommonModule],
                declarations: [MdRipple],
                providers: [VIEWPORT_RULER_PROVIDER],
            },] },
];
/**
 * @nocollapse
 */
MdRippleModule.ctorParameters = function () { return []; };
function MdRippleModule_tsickle_Closure_declarations() {
    /** @type {?} */
    MdRippleModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    MdRippleModule.ctorParameters;
}
