// ng-pwa-tools/bin/ngu-firebase-push.js
#!/usr/bin/env node_modules/.bin/ts-node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var minimist = require("minimist");
var fs = require("fs");
var firebase_push_1 = require("../lib/firebase-push");
var args_1 = require("../lib/common/args");
var util_1 = require("../lib/common/util");
var args = minimist(process.argv.slice(2), {
    string: ['module', 'base-href', 'dist', 'lazy-root', 'index', 'in', 'out'],
});
args_1.requiredArgs(args, ['dist', 'module', 'lazy-root', 'index']);
var dist = util_1.toAbsolute(args['dist']);
var appModule = util_1.toAbsolute(args['module']);
var loadChildrenRoot = util_1.toAbsolute(args['lazy-root']);
var index = util_1.toAbsolute(args['index']);
var baseHref = args['base-href'] || '/';
var config = {
    appModule: appModule,
    baseHref: baseHref,
    dist: dist,
    index: index,
    loadChildrenRoot: loadChildrenRoot,
};
if (args['in']) {
    config.config = util_1.toAbsolute(args['in']);
}
firebase_push_1.generateFbPushConfig(config)
    .then(function (config) { return JSON.stringify(config, null, 2); })
    .then(function (config) {
    if (args['out']) {
        var out = util_1.toAbsolute(args['out']);
        fs.writeFileSync(out, config);
    }
    else {
        console.log(config);
    }
});
