// @angular/material/progress-spinner/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { MdCommonModule } from '../core';
import { MdProgressSpinner, MdSpinner, MdProgressSpinnerCssMatStyler, } from './progress-spinner';
var MdProgressSpinnerModule = (function () {
    function MdProgressSpinnerModule() {
    }
    return MdProgressSpinnerModule;
}());
MdProgressSpinnerModule.decorators = [
    { type: NgModule, args: [{
                imports: [MdCommonModule],
                exports: [
                    MdProgressSpinner,
                    MdSpinner,
                    MdCommonModule,
                    MdProgressSpinnerCssMatStyler
                ],
                declarations: [
                    MdProgressSpinner,
                    MdSpinner,
                    MdProgressSpinnerCssMatStyler
                ],
            },] },
];
/**
 * @nocollapse
 */
MdProgressSpinnerModule.ctorParameters = function () { return []; };
function MdProgressSpinnerModule_tsickle_Closure_declarations() {
    /** @type {?} */
    MdProgressSpinnerModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    MdProgressSpinnerModule.ctorParameters;
}
export { MdProgressSpinnerModule };
export { PROGRESS_SPINNER_STROKE_WIDTH, MdProgressSpinnerCssMatStyler, MdProgressSpinnerBase, _MdProgressSpinnerMixinBase, MdProgressSpinner, MdSpinner } from './progress-spinner';
