// @angular/material/toolbar/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { MdCommonModule } from '../core';
import { MdToolbar, MdToolbarRow } from './toolbar';
var MdToolbarModule = (function () {
    function MdToolbarModule() {
    }
    return MdToolbarModule;
}());
export { MdToolbarModule };
MdToolbarModule.decorators = [
    { type: NgModule, args: [{
                imports: [MdCommonModule],
                exports: [MdToolbar, MdToolbarRow, MdCommonModule],
                declarations: [MdToolbar, MdToolbarRow],
            },] },
];
/**
 * @nocollapse
 */
MdToolbarModule.ctorParameters = function () { return []; };
function MdToolbarModule_tsickle_Closure_declarations() {
    /** @type {?} */
    MdToolbarModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    MdToolbarModule.ctorParameters;
}
export { MdToolbarRow, MdToolbarBase, _MdToolbarMixinBase, MdToolbar } from './toolbar';
