// @angular/material/checkbox/checkbox-required-validator.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import * as tslib_1 from "tslib";
import { Directive, forwardRef, } from '@angular/core';
import { CheckboxRequiredValidator, NG_VALIDATORS, } from '@angular/forms';
export var /** @type {?} */ _MdCheckboxRequiredValidator = CheckboxRequiredValidator;
export var /** @type {?} */ MD_CHECKBOX_REQUIRED_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(function () { return MdCheckboxRequiredValidator; }),
    multi: true
};
/**
 * Validator for Material checkbox's required attribute in template-driven checkbox.
 * Current CheckboxRequiredValidator only work with `input type=checkbox` and does not
 * work with `md-checkbox`.
 */
var MdCheckboxRequiredValidator = (function (_super) {
    tslib_1.__extends(MdCheckboxRequiredValidator, _super);
    function MdCheckboxRequiredValidator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return MdCheckboxRequiredValidator;
}(_MdCheckboxRequiredValidator));
export { MdCheckboxRequiredValidator };
MdCheckboxRequiredValidator.decorators = [
    { type: Directive, args: [{
                selector: 'md-checkbox[required][formControlName],' +
                    'mat-checkbox[required][formControlName],' +
                    'md-checkbox[required][formControl],md-checkbox[required][ngModel],' +
                    'mat-checkbox[required][formControl],mat-checkbox[required][ngModel]',
                providers: [MD_CHECKBOX_REQUIRED_VALIDATOR],
                host: { '[attr.required]': 'required ? "" : null' }
            },] },
];
/**
 * @nocollapse
 */
MdCheckboxRequiredValidator.ctorParameters = function () { return []; };
function MdCheckboxRequiredValidator_tsickle_Closure_declarations() {
    /** @type {?} */
    MdCheckboxRequiredValidator.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    MdCheckboxRequiredValidator.ctorParameters;
}
