// @angular/material/checkbox/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ObserveContentModule } from '@angular/cdk/observe-content';
import { MdRippleModule, MdCommonModule, FocusOriginMonitor } from '../core';
import { MdCheckbox } from './checkbox';
import { MdCheckboxRequiredValidator } from './checkbox-required-validator';
var MdCheckboxModule = (function () {
    function MdCheckboxModule() {
    }
    return MdCheckboxModule;
}());
export { MdCheckboxModule };
MdCheckboxModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, MdRippleModule, MdCommonModule, ObserveContentModule],
                exports: [MdCheckbox, MdCheckboxRequiredValidator, MdCommonModule],
                declarations: [MdCheckbox, MdCheckboxRequiredValidator],
                providers: [FocusOriginMonitor]
            },] },
];
/**
 * @nocollapse
 */
MdCheckboxModule.ctorParameters = function () { return []; };
function MdCheckboxModule_tsickle_Closure_declarations() {
    /** @type {?} */
    MdCheckboxModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    MdCheckboxModule.ctorParameters;
}
export { MD_CHECKBOX_CONTROL_VALUE_ACCESSOR, TransitionCheckState, MdCheckboxChange, MdCheckboxBase, _MdCheckboxMixinBase, MdCheckbox } from './checkbox';
export { _MdCheckboxRequiredValidator, MD_CHECKBOX_REQUIRED_VALIDATOR, MdCheckboxRequiredValidator } from './checkbox-required-validator';
