// @angular/material/sort/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { MdSortHeader } from './sort-header';
import { MdSort } from './sort';
import { MdSortHeaderIntl } from './sort-header-intl';
import { CommonModule } from '@angular/common';
export { MdSortHeader } from './sort-header';
export { MdSortHeaderIntl } from './sort-header-intl';
export { MdSort } from './sort';
var MdSortModule = (function () {
    function MdSortModule() {
    }
    return MdSortModule;
}());
export { MdSortModule };
MdSortModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                exports: [MdSort, MdSortHeader],
                declarations: [MdSort, MdSortHeader],
                providers: [MdSortHeaderIntl]
            },] },
];
/**
 * @nocollapse
 */
MdSortModule.ctorParameters = function () { return []; };
function MdSortModule_tsickle_Closure_declarations() {
    /** @type {?} */
    MdSortModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    MdSortModule.ctorParameters;
}
